plugins/ssr/ssrview.ml

```diff
@@ -305,51 +319,78 @@ end
 
 let pose_proof subject_name p =
   Tactics.generalize [p] <*>
-  Option.cata
-    (fun id -> Ssrcommon.tclRENAME_HD_PROD (Name.Name id)) (tclUNIT())
-    subject_name
+  begin match subject_name with
+  | id :: _ -> Ssrcommon.tclRENAME_HD_PROD (Name.Name id)
+  | _ -> tclUNIT() end
   <*>
   Tactics.New.reduce_after_refine
 
-let rec apply_all_views ~clear_if_id ending vs s0 =
+let rec apply_all_views_aux ~clear_if_id vs finalization conclusion s0 =
   match vs with
-  | [] -> ending s0
+  | [] -> finalization s0 (fun name p ->
+            match p with
+            | None -> conclusion ~to_clear:[]
+            | Some p ->
+                pose_proof name p <*> conclusion ~to_clear:name)
   | v :: vs ->
       Ssrprinters.ppdebug (lazy Pp.(str"piling..."));
-      is_tac_in_term v >>= function
-      | `Tac tac ->
-           Ssrprinters.ppdebug (lazy Pp.(str"..a tactic"));
-           ending s0 <*> Tacinterp.eval_tactic tac <*>
-           Ssrcommon.tacSIGMA >>= apply_all_views ~clear_if_id ending vs
+      is_tac_in_term ~extra_scope:"ssripat" v >>= function
       | `Term v ->
            Ssrprinters.ppdebug (lazy Pp.(str"..a term"));
            pile_up_view ~clear_if_id v <*>
-           apply_all_views ~clear_if_id ending vs s0
+           apply_all_views_aux ~clear_if_id vs finalization conclusion s0
+      | `Tac tac ->
```

Comment:
Can you write a comment explaining what we do in the `Tac` case?

Reply:
End the application of Term views accumulated so far, apply tac, continue. I'd like to factor more the common code (see the empty list case) but I was not able